integration/kotlinx-coroutines-firebase-android/kotlinx-coroutines-firebase-android/src/main/java/kotlinx/coroutines/experimental/firebase/android/DatabaseReference.kt

```diff
@@ -0,0 +1,266 @@
+                continuation.resumeWithException(exception)
+            }
+        }
+    }
+
+    continuation.invokeOnCompletion { reference.removeEventListener(listener) }
+
+    reference.addListenerForSingleValueEvent(listener)
+}
+
+/**
+ * Coroutine to read a single value from Firebase Database
+ *
+ * Example:
+ * ```kotlin
+ * val message = FirebaseDatabase.getInstance()
+ *     .getReference("message")
+ *     .readValue(String::class.java)
+ *
+ * println(message) // Hello World
+ * ```
+ *
+ * @param type Expected type wrapped in a Class instance.
+ * @param T The type of the expected value from Firebase Database.
+ * @return The persisted object of the type T informed.
+ */
+suspend fun <T : Any> DatabaseReference.readValue(type: Class<T>): T = readReference(this, type)
+
+/**
+ * Coroutine to read a single value from Firebase Database
+ *
+ * ```kotlin
+ * val message = FirebaseDatabase.getInstance()
+ *     .getReference("message")
+ *     .readValue<String>()
+ *
+ * println(message) // Hello World
+ * ```
+ *
+ * @param T The type of the expected value from Firebase Database.
+ * @return The persisted object of the type T informed.
+ */
+suspend inline fun <reified T : Any> DatabaseReference.readValue(): T = readValue(T::class.java)
+
+/**
+ * Coroutine to read a collection of values from Firebase Database
+ *
+ * This method is not intended to be used in the code. It's just a base function that contains the
+ * main coroutine code that reads a collection of values from Firebase Database. To use it's
+ * functionality call [kotlinx.coroutines.experimental.firebase.android.readValue] function.
+ *
+ * The implementation consists of a [suspendCancellableCoroutine] that encapsulates a
+ * [ValueEventListener].
+ *
+ * @param reference The Firebase Database node to be read.
+ * @param type Expected type wrapped in a Class instance.
+ * @param T The type of the expected value from Firebase Database.
+ * @return The persisted object of the type T informed.
+ */
+private suspend fun <T : Any> readReferences(
+        reference: DatabaseReference,
+        type: Class<T>
+): List<T> = suspendCancellableCoroutine { continuation ->
+    val listener = object : ValueEventListener {
+
+        /**
+         * Callback to handle Firebase Database errors
+         *
+         * It's triggered when there are authentication problems or network error. To see all
+         * possible, pleese refer to Firebase Database docs.
+         *
+         * @link https://firebase.google.com/docs/reference/android/com/google/firebase/database/DatabaseError
+         * @param error The error occurred.
+         * @throws FirebaseException When the error happens within Firebase SDK.
+         */
+        override fun onCancelled(error: DatabaseError) {
+            val exception = when (error.toException()) {
+                is FirebaseException -> error.toException()
+                else -> Exception("The Firebase call was canceled")
```

Comment:
This is a bad practice. Don't omit information about errors when propagating it.
Otherwise, someone will have "The Firebase call was canceled" in logs and nothing else.

Also, an exception is created twice via `toException`, please do it once.

Same about second `onCancelled`

Reply:
Changed in https://github.com/Kotlin/kotlinx.coroutines/pull/426/commits/833cdb9e37b1da841f7a9eba68a7b02021c6b738.